### DISCUSSION TITLE
chapter_data_structure/classification_of_data_structure/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#201] [Announcements]
# chapter_data_structure/classification_of_data_structure/

一本动画图解、能运行、可提问的数据结构与算法入门书

https://www.hello-algo.com/chapter_data_structure/classification_of_data_structure/

**Comment by DesertFishing** [comment]
请问：线性数据结构和非线性数据结构，哈希表为什么都存在。

**Reply by chenzhiwei54czw** [reply]
我觉得应该是因为哈希表在没有冲突之前，是用数组存储的，当出现哈希冲突之后，有各种解决冲突的方式，比如链地址法，所以两种逻辑结构都包括它，不像其他的数据结构都是只用数组或者只用链表实现的

**Reply by krahets** [reply]
好问题～这因为哈希表底层是数组，所以包含线性数据结构。而为了解决哈希冲突，我们可能会使用“链式地址”（后续散列表章节会讲）。在链式地址中，哈希表的每个桶（索引）都存储一个链表；当链表长度超过一定阈值时，又可能被转化为树（常为红黑树），以提升查询效率。因此，哈希表可能同时包含线性（数组、链表）和非线性（树）数据结构。

**Reply by SixAnChen** [reply]
请问数据结构有没有推荐学习的书籍或者博客，比较通俗易懂一点的

**Reply by HighBigSky** [reply]
哈希表在jdk1.8之后就是数组链表+红黑树了，在长度大于8之后就会树化，爪哇

**Comment by TujinLee** [comment]
基于数组实现的数据结构也被称为「静态数据结构」，这意味着该数据结构在在被初始化后，长度不可变。

像JS的数组可以不断的push数据，长度也在变化吧，可能底层做了处理？

**Reply by krahets** [reply]
是的 Python List, Go Slice, JS/TS Array, C++ vector 都采用了类似的策略，用户最常用的是动态数组，底层实现了扩容机制。（语言没列全，现在一般都直接用动态数组了，方便）

**Comment by luckyamber1992** [comment]
2023.05转码打卡，看晕了

**Reply by krahets** [reply]
万事开头难，加油！

**Comment by Sakura11hoo** [comment]
算法小白，冲今年的秋招

**Reply by YuMiao316** [reply]
ヾ(◍°∇°◍)ﾉﾞ加油

**Reply by Tttdotd** [reply]
加油，你绝对没问题的

**Comment by chusifeng** [comment]
「物理结构」反映了数据在计算机内存中的存储方式，可分为数组的连续空间存储和链表的离散空间存储
---- 这个是指程序“看到的”虚拟内存是连续空间还是说虚拟映射的物理都是连续空间？

**Reply by krahets** [reply]
好问题！这里指的是虚拟内存。假设我们初始化一个很大的数组。在虚拟内存中，数组的内存空间是连续的。这样我们才能通过首元素地址加偏移量的方式来访问数组中的任何元素。

然而，在物理内存中，数组可能并不连续。虚拟内存到物理内存的映射涉及到分页（paging）技术。操作系统通过分页系统，将虚拟内存和物理内存都划分为大小相等的页。每个虚拟页都可以单独映射到物理内存中的任意一页。所以，一个大的数组（占据多个虚拟内存页）在物理内存中可能是分散的，每个虚拟页在物理内存中的位置可能不连续。

值得强调的是，虽然物理内存中的连续性可能会对性能有影响，但在实际应用中，这通常并不是主要的性能瓶颈。现代的计算机系统已经采用了许多优化技术，例如缓存、预取等，以尽可能减小这种影响。

**Comment by YuMiao316** [comment]
作者的链表的图看着很舒服，数据域和指针域感觉很清晰

**Comment by zhuasdfg** [comment]
节点和节点指针看着一目了然，链表就该长这样

**Comment by 1ess** [comment]
Fig. 线性与非线性数据结构中队列的操作称为enqueue和dequeue比较合适吧

**Reply by krahets** [reply]
只有 C# 的入队出队方法是用 `enqueue` 和 `dequeue` 命名的。感觉这个不需要太纠结。

**Comment by 1ess** [comment]
上文中哈希表被描述为既是线性结构还是树形结构，我认为其属于线性结构比较合理把

**Reply by 1ess** [reply]
看了后面的QA，能理解想表达的意思，但是感觉这块描述还是不合理，这个地方还是应该要描述清楚。线性结构：...哈希表，元素之间是一对一的顺序关系。树形结构：... 哈希表，元素之间是一对多的关系。

**Reply by krahets** [reply]
从存储的角度来看，哈希表的底层是线性的数组结构。这个数组中的每一个槽位可能包含一个值，也可能包含一个链表或树（在解决哈希冲突时），因此也可能包含非线性数据结构。

**Comment by 1ess** [comment]
Fig. 连续空间存储与离散空间存储中，链表箭头的起始位置和结束位置正确吗，我理解不应该是起始位置为指针域的中心，结束位置是下一个节点的数据域吗？

**Comment by iHaveFiveDogs** [comment]
图做的确实有新意。

**Comment by wangxiangnan** [comment]
逻辑结构和物理结构都有数组和链表的概念，这个如何区分呢？之前是通过顺序存储和链式存储区分的。

**Reply by krahets** [reply]
在逻辑结构中，数组和链表都属于线性结构。在物理结构中，数组是顺序存储（连续内存空间），链表是链式存储（离散内存空间）

**Comment by BenjiWu** [comment]
线性结构：数组、链表、队列、栈、哈希表，元素之间是一对一的顺序关系。
树形结构：树、堆、哈希表，元素之间是一对多的关系。
这里hash table出现在两个分类，是写错了还是都符合呢？

**Reply by krahets** [reply]
这个在小结的 Q&A 里有讲

**Reply by newcakes** [reply]
这让我回想起来c++中map容器和multimap容器的区别，map容器的key值不可以重复，而multimap容器是可以重叠的，这两个容器的底层应该是hash table，这会不会就是一对一和一对多的区别呢？@krahets

**Reply by krahets** [reply]
@newcakes 这里确实有一对一和一对多的逻辑关系。即使是一对一的普通哈希表，它的底层实现也可能包含非线性数据结构。

**Reply by yicijohncy** [reply]
遇到一样的疑问，hhhh，k 大，建议把Q&A那部分内容挪到分类讲解下方，也不会卡住，能理解的~

**Comment by Chenyuyang7** [comment]
大佬你好，在java中ArrayList底层采用hash表的存储方式，在jdk7之前hash表存储数据采用的是的数组加链表，jdk7之后采用数组加链表及红黑树的存储方式，那么我的问题是，hash表本身提供了这种数组加链表和红黑树的存储方式还是这种存储方式是java所设计的？

**Reply by krahets** [reply]
这种存储方式是 Java 设计的。哈希表是使用数组、链表、二叉树等这些数据结构实现的。

**Comment by Zswdhy** [comment]
基于数组实现的数据结构也被称为“静态数据结构”
这个存在歧义吧，栈，队列也是可以进行出栈、入栈等操作的呀

**Reply by krahets** [reply]
Hi，栈和队列确实可以实现动态的数据操作，但数据结构仍然是“静态”（长度不可变）的。尽管基于数组的数据结构可以动态地添加或删除元素，但它们的容量是固定的。如果数据量超出了预分配的大小，就需要创建一个新的更大的数组，并将老数组的内容复制到新数组中。

**Reply by Zswdhy** [reply]
在构建栈/队列的时候，未指定栈/队列的大小，怎么理解成“静态数据结构呢”

**Reply by krahets** [reply]
在高级编程语言中，我们无须人工指定栈/队列的初始容量，这个工作是由类内部自动完成的。例如，Java 的 ArrayList 的初始容量通常为 10 。另外，扩容操作也是自动实现的。详见本书的“列表”章节。

也就是说，基于数组实现的栈，底层实现是数组，仍然具有数组的静态性。只不过编程语言的内置库已经预先实现好了栈类的各种行为（初始化、扩容等），使得它看起来是动态的，使用起来是灵活的。

**Reply by LitchiLiy** [reply]
我记得C++容器好像未指定容量大小则默认为0, 当被添加元素时会将存储空间搬到其他地方并赋予一个合适的容量, 等容量再次被超出时会再次将存储空间搬到其他地方并再次赋予一个更大的容量.

**Comment by clearwater753** [comment]
“物理结构从底层决定了数据的访问、更新、增删等操作方法，同时在时间效率和空间效率方面呈现出互补的特点”
兄弟你好，这句话中的互补是什么意思

**Reply by PgSheep615** [reply]
时间换空间，空间换时间。

**Reply by Harbour-z** [reply]
牺牲时间效率提高空间效率，或者牺牲空间效率换时间效率

**Comment by Harbour-z** [comment]
* 所有数据结构都是基于数组（静态）、链表（动态）或者他们的混合实现的

**Comment by lianlilin** [comment]
物理结构从底层决定了数据的访问、更新、增删等操作方法，同时在时间效率和空间效率方面呈现出互补的特点

这句话说的很好啊！

**Comment by DarKnightL** [comment]
Stack and Queue

Stack:
Principle: LIFO (Last In, First Out) - The last item to be inserted into a stack is the first one to be deleted from it.
Real-world analogy: Think of a stack like a stack of plates on a table. The last plate you put on the top of the stack is the first plate you would remove if you were unstacking them.
Operations:
Push: Adds an item onto the stack. If the stack is full, then it is considered a stack overflow condition.
Pop: Removes an item from the stack. Items are removed in the reverse order of their addition (LIFO). If the stack is empty, trying to pop an element could lead to a stack underflow condition.
Usage: Stacks are used in scenarios like function calls (call stacks), undo mechanisms in text editors, and for depth-first search algorithms in data structures like trees or graphs.


Queue:
Principle: FIFO (First In, First Out) - The first item to be inserted into a queue is the first one to be deleted from it.
Real-world analogy: Think of a queue as a line of people waiting to buy tickets at a ticket counter. The person who has been in line the longest (first in) gets served first (first out).
Operations:
Enqueue: Adds an item to the end of the queue. If the queue is full, then it is considered a queue overflow condition.
Dequeue: Removes an item from the front of the queue. If the queue is empty, trying to dequeue an element could lead to a queue underflow condition.
Usage: Queues are used in scenarios like breadth-first search algorithms in data structures like trees or graphs, managing tasks in printers, process scheduling, and in simulations of real-world systems (like customers standing in lines).
In summary, the main difference lies in the order of element retrieval: stacks retrieve elements in the reverse order of their addition (LIFO), while queues retrieve elements in the same order as their addition (FIFO). This fundamental characteristic makes them suitable for different kinds of operations in computing and programming.

**Comment by Xj1567589354** [comment]
daka

**Comment by lizhuangShen** [comment]
逻辑结构：
	线性：数组，链表，栈，队列，哈希表
	非线性：树，图，堆，哈希表


线性结构：元素之间是一对一的顺序关系	数组，链表，栈，队列，哈希表
树形结构：元素之间值一对多的关系		树，堆，哈希表
网状结构：元素之间是多对多的关系		图

物理结构：
	所有的数据结构都是基于数组或链表实现的

基于数组：栈，队列，哈希表，树，图，堆，矩阵，张量（维度>=3的数组）
基于链表：栈，队列，哈希表，树，图

**Comment by lihuangshuaige** [comment]
堆和树有啥区别？

**Reply by krahets** [reply]
请看[这里](https://www.hello-algo.com/chapter_heap/heap/)

**Comment by java-pakchoi** [comment]
#### 基于数组可实现：栈、队列、哈希表、树、堆、图、矩阵、张量（维度>=3的数组）等
 请问一下，数组不是线性结构嘛，怎么实现非线性的树呢？
 难道是多维数组或者借助其他引用类型？

**Reply by hemingguo** [reply]
应该取决于具体的树的类型，比如二叉树，前缀树就可以用数组实现

**Reply by QiuShui-666** [reply]
大顶堆也是树的一种，堆那块就是用数组实现的

**Reply by krahets** [reply]
Hi，在本书的[这个章节](https://www.hello-algo.com/chapter_tree/array_representation_of_tree/)有讲

**Comment by brycenhe** [comment]
请问文中的配图是使用什么软件制作的呢？感觉很棒！

**Reply by krahets** [reply]
谢谢～Keynote 或者 Powerpoint 都可以画

**Comment by giftfm** [comment]
线性结构：数组、链表、队列、栈、哈希表，元素之间是一对一的顺序关系。
树形结构：树、堆、哈希表，元素之间是一对多的关系。 
hash表为什么既可以一对一又可以多对一，hash表怎么多对一？

**Reply by ganxingrui** [reply]
按map理解，一对一就是key-value;一对多，把value换成链表

**Reply by krahets** [reply]
Hi，一对多的关系与“哈希冲突”概念相关。在链式地址哈希表中，冲突的元素被存储在一个链表中，这条链表还可能被优化为二叉树，以提升查找效率。

**Reply by HannahHathaway** [reply]
哦哦 明白啦 非常感谢

**Comment by ganxingrui** [comment]
逻辑结构：是抽象结构，理论上的结构；
物理结构：是存储在物理硬件上的真实结构；

**Comment by foreverpersistence** [comment]
逻辑结构： 元素逻辑关系，分线性和非线形。 
 - 线形结构： 数组、链表、堆、栈、哈希， 元素 一对一
 - 非线形结构：树、图、哈希
物理结构：连续 和 非连续
-  连续结构： 数组结构
- 非连续结构： 链表结构

所有的数据结构可以通过 数组和链表实现。

作者讲的很形象，赞

**Comment by LoganQiu** [comment]
k神，想问问这些图片是用什么工具制作的，看起来非常优雅

**Reply by krahets** [reply]
用 keynote 或 powerpoint 都可以画噢～

**Reply by HuaWang135608** [reply]
可以尝试一下 `draw.io` 哇

**Comment by CryusXin** [comment]
基于数组实现的数据结构也称“静态数据结构”，这意味着此类数据结构在初始化后长度不可变。

但是python中的数组通常是指list，这是一个动态数据结构，然后python中还有一个array模块，提供了一种数组类型
````python
import array

# 创建一个整数数组（'i'是类型码，表示有符号整数）
int_array = array.array('i', [1, 2, 3, 4, 5])
````
array模块的数组与列表类似，也是动态的，可以在运行时改变大小

python中只有元组类型是静态数据结构

**Reply by krahets** [reply]
Hi，感谢反馈！文中的描述的确有歧义，我想想怎么改。

你提到的元祖静态、列表动态是从“逻辑结构”角度观察的。其实，如果从逻辑角度看，数组也可以是“动态”的：使用一个变量记录有效容量，我们就可以模拟出“动态性”了。

然而，文中提到的“静态”和“动态”概念，是从“物理结构”角度观察的。虽然列表（list）可以进行增删改查，但由于它的底层是数组，因此在扩容过程中仍然需要进行内存再分配、元素整体搬移的额外工作，即它的物理结构是“静态”的。

这个名词从不同角度观察会有不同的结果（我尚未找到标准定义，后面我再看看）。

**Reply by CryusXin** [reply]
我进行了一些简单的尝试
````python
nums = [1, 2, 3, 4, 5]
print(f'nums内存地址：{hex(id(nums))}')
nums.insert(1, 6)
print(nums)
print(f'nums插入元素后的内存地址：{hex(id(nums))}')
````
输出
nums内存地址：0x24e220d9480
[1, 6, 2, 3, 4, 5]
nums插入元素后的内存地址：0x24e220d9480

这个插入操作并没有改变它的内存地址，如果说数组扩容的时候会进行内存再分配，那我的理解，只有一种可能，就是python列表的插入删除方法，实际上用空间换时间了，在我定义这个nums的时候，就已经默认可以插入100个或者更多的元素，当我插入的元素没有超过默认值的时候，其实并不需要进行内存再分配，只有空间不足以容纳新元素时，python的解释器才会去获取更大的内存空间，然后执行元素复制

**Reply by krahets** [reply]
是的，扩容机制在“列表”章节有介绍：系统会为数组初始化一个合理的长度，在到达这个长度限制之前，添加元素不会触发扩容；当超出长度限制时，则会进行扩容（内存再分配 + 元素搬移）。

由于 `nums` 是一个对象，因此即使扩容之后 `id(nums)` 输出的地址应该也不会改变（底层存储数据的数组的内存地址会发生变化）。

**Comment by yxsongbo** [comment]
从图中看不出堆和树的区别嘛

**Reply by anchorforG** [reply]
因为这个排版吧，理解为是则不是的状态。阅读后面内容可以消除了，只是没找到怎么删除这个评论了。

**Reply by krahets** [reply]
Hi，这是因为堆是一种树的特例情况

**Comment by NgxuAnGMH** [comment]
编程语言的标准包，都会将常用的数据结构封装好提供外部使用。虽然“数组”作为“静态数据结构”，但其实在一些编程语言中也已经考虑到了动态伸缩的设计。不过掌握该知识点还是很有必要的

- 数组可通过重新分配内存实现长度变化，从而具备一定的“动态性”。

**Comment by johnluck66** [comment]
您好！我有几个问题，请赐教：
问题1：
书中提到："当算法程序运行时，正在处理的数据主要存储在内存中。"  ，我联想：
请问在CPU内部的寄存器、缓存，请问 是不是也有有 书中的 数据结构，和用到相关的算法？
这方面 有什么不一样吗？

问题2：
书中提到：“计算机根据特定规则为表格中的每个单元格分配编号，确保每个内存空间都有唯一的内存地址。”
我联想：
请问 （1）这个“特点规则” 有没有行业统一标准？
请问（2） 不同的操作系统、不同CPU、不同内存条，这个“特点规则” 是否不同？
（3）在电脑端的和手机移动端， 这个“特点规则” 有什么不一样？

问题3：
书中提到”所有数据结构都是基于数组、链表或二者的组合实现的“
我理解：
1、数组、链表 是物理存储的 两个最基本形态，即 连续 和 分散，也只有 这两种基本形态
2、其他所有 数据结构，是 根据 现实世界的需求（或者说 现实世界的问题 的抽象 及分类），人为创造出 各种数据结构，  像 初级语言 向 高级语言 的演进
请问我的理解是否对，请赐教

问题4：
 书中”图3-3“ 中，存储链表，”存储节点值“ （横向）后面紧接 ”存储节点指针“，
请问（1）是不是所有（操作系统、CPU、内存） 中的 链表，都是这种设置？
（2）箭头 指向 的好像是 节点值 和节点指针 之间，请问这个是不是可以展开讲一讲?

**Reply by BLackWatch1118** [reply]
1：我的理解是算法实际上是一种思想，数据结构是实现你思想的载体，不同地方由于目的不同适用的思想也不同，比如缓存替换中可能有随机、FIFO、LRU之类的替换方法，你想实现这种方法就要选择合适的数据结构，比如在需要取得某链表节点的前一个节点时用单纯的单向链表就比较困难了。

2：这个比较复杂，我不是很确定，因为实际上内存机制包括虚拟内存和实际内存和缓存，虚拟内存提供一个对主存的抽象，让每一个进程都认为它享用一个巨大、一致、私有的地址空间，缓存是加速数据传输的中间层。cpu生成的虚拟地址通过地址翻译单元MMU来转换为物理地址，就是在主存上数据的实际储存位置，然后具体的过程中会有其他的机制来保证这个过程正确且高效的执行。我觉得这个是跟着操作系统走的，基本思路应该类似，但不同的操作系统的实现细节应该不太一样，32位和64位的系统起码地址空间就不一样。

3：不知道。我觉得都是为了解决问题找出的解决办法，需要啥用啥，我不太清楚初级和高级的定义。

4：仍然我按照之前想法，链表也是一种思路，核心就是每个节点都有找到下一个节点的方法，所以实现思路就是节点本身的值+到下一个节点的索引，至于这个索引是节点的指针还是其他标识方式都无所谓，能找到就行，看编程语言是怎么实现的。C语言里应该是指向下一个节点首元素的地址，图里应该只是为了看起来好看，指向中间可能会因为对齐之类的原因取值比较麻烦。

不确定我说的是不是对的，你可以多方求证下。

**Reply by oopfjp** [reply]
关于第四个问题的第二小点，鄙人是这么理解的：链表（这里举例单链表）的每个元素都包含一个值和一个指向下一个元素的指针，所以箭头就代表Next指针，指的就是下一个元素的内存地址，你说的指向节点值和节点指针中间，可能就是这么画的，只要理解原理就行了。

**Comment by lvmingq** [comment]
书中怎么没有提到线性表

**Reply by krahets** [reply]
线性表是一个中文社区特有的概念，它的英文翻译 Linear list 在英文社区并不常见，因此本书没有介绍这个概念。

详见 Wiki，线性表的词条只有中文版本：
https://zh.wikipedia.org/wiki/%E7%BA%BF%E6%80%A7%E8%A1%A8

相对地，数组、链表、列表是中英文社区（以及其他语言）都广泛接受的概念。

**Comment by fr33bug** [comment]
内存是所有程序的共享资源，当某块内存被某个程序占用时，则无法被其他程序同时使用了。

-- 这个说法并不是太严谨，通常情况下是这样，但也存在共享内存的情况。

**Reply by lei-ctyh** [reply]
作者应该在文中说明了     "通常无法被其他程序同时使用了。"

**Comment by BeOneBetterGUO** [comment]
为什么线性结构和非线性结构中都有哈希表？？？

**Reply by BeOneBetterGUO** [reply]
omg我看见了回复了，这条评论能不能删掉？

**Reply by xchsh12345** [reply]
哈希表的实现可以同时包含线性和非线性结构。

**Comment by 2oTA** [comment]
可惜没早点看到这个网站

**Comment by AritxOnly** [comment]
有的时候上课会提到的：**堆区**和**栈区**这个模型是否就是完全对应对应物理结构的连续与分散？还是可能有什么因为环境导致的区别？

**Comment by Umizhang0910** [comment]
henbuduo

**Comment by Stev1eL** [comment]
王长生，看到我没

**Comment by rui-project** [comment]
Conclusion：
线性数据结构：数组、链表、栈（先进后出）、队列（先进先出）、哈希表（一对一）；
非线性数据结构：树、堆、哈希表（一对多），图（多对多）；

栈（stack）：先进后出；
队列（queue）：先进先出；
堆：也被称为“优先队列”（priority queue）；

所有数据结构都是基于数组、链表或二者的组合实现的：
基于数组可实现：栈、队列、哈希表、树、堆、图、矩阵、张量（维度 >=3 的数组）等;
基于链表可实现：栈、队列、哈希表、树、堆、图等;

**Comment by HighBigSky** [comment]
数组默认初始长度为10，后面1.5倍扩容，爪哇

**Comment by WhitGiv** [comment]
day2

**Comment by vuvivian** [comment]
25.03.24： 队列那个图感觉和 栈保持同一方向的 push 更容易记忆

**Comment by webLiang** [comment]
牛批

**Comment by sunny20010115** [comment]
第二天打卡

**Comment by OptimusHimself** [comment]
写得真好哇。从学校学完第一遍，再来看作者的总结，醍醐灌顶。

**Comment by zhunaoguozai1** [comment]
这个哈希表存在1对1顺序，所以满足线性，也存在一对多所以又满足非线性是吗

**Reply by wumath015-max** [reply]
补充明确一下线性结构和非线性结构的概念吧。
线性结构：元素之间是 “一对一” 的关系，非线性结构：元素之间存在 “一对多” 的关系。
所以理解应该是正确的。

**Comment by Changkeyii** [comment]
day2

**Comment by L111M1** [comment]
学数据结构其实并不是说一定得用，其实真正学数据结构是为了了解分散与连续的内存思想。

**Comment by changwang123** [comment]
打卡！

**Comment by Modellife** [comment]
打卡

**Comment by illusionaryDaytime-l** [comment]
数组为什么会是“静态数据结构”呢？不是可以通过append、pop等方法对数组长度进行改变吗

**Reply by t0ugh-sys** [reply]
python的list底层实现是基于静态数组，如果当前分配的空间够，就直接放进去。如果空间不够，Python 会自动重新分配一个更大的数组，然后把旧元素拷贝到新数组。

**Comment by Luckylvbingwen** [comment]
哈希表既是线性数据结构又是非线性数据结构吗

**Reply by fynixw** [reply]
我猜是和哈希表具体实现有关，使用开放寻址的哈希表实现是线性结构，使用红黑树解冲突的哈希实现算是非线性结构吧。

**Reply by QinQG** [reply]
确实应该备注一个“包含哈希冲突”

**Reply by DeanYao2014** [reply]
hash table的物理结构是线性的吧 ,但是逻辑结构不是, 这里即使线性又是非线性的确容易搞混.

**Comment by zhuke-ai** [comment]
1. 内存峰值与系统稳定性（容量维度）
“算法所占用的内存峰值不应超过系统剩余空闲内存”
什么是内存峰值？
算法在运行过程中，某一时刻占用的内存最大值。例如，一个排序算法在处理数据时，可能需要申请一个临时数组来存放中间结果，此时它的内存占用达到最高点。[1][2]
为什么要控制峰值？
防止崩溃（OOM）： 如果峰值超过了系统剩余的物理内存，操作系统会抛出“内存溢出”（Out of Memory）错误，直接导致程序强制退出或崩溃。[1][2]
避免性能骤降（Swapping）： 当物理内存不足时，系统会将部分内存数据写入磁盘（虚拟内存/交换分区）。磁盘的读写速度比内存慢数千倍，这会导致算法运行极其缓慢。[1][2]
设计原则： 算法设计者不仅要看平均内存占用，更要关注“最坏情况”下的最大占用量，确保在硬件红线内运行。[1][2]
2. 连续空间 vs. 分散空间（物理维度）
“如果缺少连续大块的内存空间，那么所选用的数据结构必须能够存储在分散的内存空间内”[1][2]
这里涉及计算机底层内存分配的一个痛点——内存碎片（Memory Fragmentation）。[1][3][4][5]
连续内存空间（对应“数组”）：
数组（Array）要求内存空间必须是连续的。[1][2][6][7]
例子： 系统虽然还剩 10MB 空闲内存，但分散成 10 个 1MB 的小块。如果你想申请一个 5MB 的数组，分配会失败，因为找不到“连续”的 5MB 空间。[1][2]
分散内存空间（对应“链表”等）：
链表（Linked List）、树（Tree）或图（Graph）这类数据结构是由一个个独立的“节点”组成的，节点之间通过指针（引用）连接。[1][2][7]
原理： 它们的节点可以散落在内存的任何角落。只要系统总空闲空间足够，哪怕没有一大块完整的内存，程序也能通过一个个小块拼凑出所需的数据存储空间。[1][2][8]
总结与应用
这段话实际上是在教你如何根据硬件现状选择合适的数据结构：
情况	内存现状	建议选用的数据结构类型	原因
情况 A	内存充足，有大块连续空间	数组（Array）	访问速度快（随机访问 
O
(
1
)
O(1)
），缓存利用率高。[1][2]
情况 B	内存碎片严重，空闲空间零散	链表（Linked List）	灵活利用碎片空间，不需要预先申请整块连续内存。[1]
通俗的比喻：
数组就像“联排别墅”，必须有一整块地皮才能盖。[1]
链表就像“分布式公寓”，只要城里哪儿有空房间（散乱的内存），租客（数据节点）就能住进去，大家通过电话（指针）互相联系即可。[1][2]
一句话总结： 高质量的算法不仅要跑得快（时间复杂度低），还得在目标机器的内存限制内“稳稳地跑”（空间复杂度适配且能容忍碎片）

**Comment by xingbaichu** [comment]
数据结构：数组、链表、栈、队列、哈希表、树、堆、图



数据结构按两个维度分：1.逻辑结构  2.物理结构

    逻辑结构：数据元素之间的逻辑关系。
    
    想要知道逻辑结构，得知道什么是逻辑关系。
    
    逻辑结构=>逻辑关系=>暂不知道

1.逻辑结构：

    1.线性结构=>数据在逻辑关系上呈线性排列=>逻辑关系
    
    2.非线性<=>不是线性结构=>线性结构
        树形结构：树，堆、哈希表
        网状结构：图
        如何区分？=>判断元素之间的关系：if(一对多) 树形 else if(多对多) 图

2.物理结构：

（1）内存

    计算机内存条上黑方块=>内存=>内存空间=>比作excel表格=>表格可以存储数据
    
 访问内存

    系统访问数据=>访问内存=>内存地址=>数据
    
    内存地址=>特定规则为表格中的每个单元格分配编号=>内存空间都有唯一的内存地址

 内存特点

    所有程序共享资源,意味着=>内存被某个程序占用时，则通常无法被其他程序同时使用

    数据结构与算法的设计，考察因素：内存资源

（2）连续与分散

    计算机内存存储方式=>物理结构=>连续空间存储(数组)
                              分散空间存储(链表)
    物理结构从底层决定=>数据访问，更新，增删等操作

连续与分散特点

     数据结构实现=>数组,链表或{数组,链表}结合
     链表实现=>程序运行过程，可对其长度进行调整<=>“动态数据结构”
     数组实现=>长度不可变<=>“静态数据结构”=>并非真不可变
